app/src/main/java/jp/yhonda/MaximaOnAndroidActivity.java

```diff
@@ -860,4 +887,82 @@ private void sessionMenu(String cmd) {
 		editText.dispatchKeyEvent(new KeyEvent(KeyEvent.ACTION_UP,
 				KeyEvent.KEYCODE_ENTER));
 	}
+
+	private void selectScriptFile() {
+		if(Build.VERSION.SDK_INT < 19) {
+			//Versions earlier than KitKat do not support the Storage Access Framework.
+			//Show file path input box instead:
+			android.support.v7.app.AlertDialog.Builder builder = new android.support.v7.app.AlertDialog.Builder(MaximaOnAndroidActivity.this);
+			LayoutInflater inflater = this.getLayoutInflater();
+			final View layout = inflater.inflate(R.layout.scriptpathalert, null);
+			builder.setView(layout);
+			final AutoCompleteTextView textView = (AutoCompleteTextView) layout.findViewById(R.id.scriptPathInput);
+			builder.setPositiveButton(R.string.ok, new DialogInterface.OnClickListener() {
+				@Override
+				public void onClick(DialogInterface dialog, int which) {
+					dialog.dismiss();
+					if(!textView.getText().toString().matches("")) {
+						copyScriptFileToInputArea(Uri.fromFile(new File(textView.getText().toString())));
+					}
+				}
+			});
+			android.support.v7.app.AlertDialog dialog = builder.create();
+			dialog.show();
+
+		} else {
+			Intent intent = new Intent(Intent.ACTION_OPEN_DOCUMENT);
+			intent.addCategory(intent.CATEGORY_OPENABLE);
+			intent.setType("*/*");
+			startActivityForResult(intent, READ_REQUEST_CODE);
+		}
+	}
+
+	private void copyScriptFileToInputArea(Uri fileUri) {
+        //Copy script file contents into a temporary file:
+        File temporaryDirectory = getApplicationContext().getCacheDir();
+        File temporaryFile = null;
+        try {
+            InputStream inputStream = getContentResolver().openInputStream(fileUri);
+            BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(inputStream));
+            String scriptLine;
+            temporaryFile = File.createTempFile("userscript", ".mac", temporaryDirectory);
+            OutputStream outputStream = getContentResolver().openOutputStream(Uri.fromFile(temporaryFile));
+            BufferedWriter bufferedWriter = new BufferedWriter(new OutputStreamWriter(outputStream));
+            long writtenBytes = 0;
+            while((scriptLine = bufferedReader.readLine()) != null) {
+                writtenBytes += (scriptLine.length() + 1);
+                if (writtenBytes > 1E7) { //Limit the size of scripts to 10MB.
```

Comment:
I got your reason. Let's limit the file size. Please define constant for 1E7. Uri can be converted to FileChannel as above. Then you can use size() method to know the size of the file.